nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/PublishMQTT.java

```diff
@@ -0,0 +1,288 @@
+        this.relationships = Collections.unmodifiableSet(relationships);
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return this.relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return descriptors;
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        try {
+            broker = context.getProperty(PROP_BROKER_URI).getValue();
+            String clientID = context.getProperty(PROP_CLIENTID).getValue();
+
+            connOpts = new MqttConnectOptions();
+            connOpts.setCleanSession(true);
+
+            PropertyValue sslProp = context.getProperty(PROP_SSL_CONTEXT_SERVICE);
+            if (sslProp.isSet()) {
+                Properties sslProps = transformSSLContextService((SSLContextService) sslProp.asControllerService());
+                connOpts.setSSLProperties(sslProps);
+            }
+
+            PropertyValue lastWillTopicProp = context.getProperty(PROP_LAST_WILL_TOPIC);
+            if (lastWillTopicProp.isSet()){
+                PropertyValue lastWillMessage = context.getProperty(PROP_LAST_WILL_TOPIC);
+                PropertyValue lastWillRetain = context.getProperty(PROP_LAST_WILL_TOPIC);
+                connOpts.setWill(lastWillTopicProp.getValue(), lastWillMessage.getValue().getBytes(), LAST_WILL_QOS, lastWillRetain.isSet() ? lastWillRetain.asBoolean() : false);
+            }
+
+            PropertyValue usernameProp = context.getProperty(PROP_USERNAME);
+            if(usernameProp.isSet()) {
+                connOpts.setUserName(usernameProp.getValue());
+                connOpts.setPassword(context.getProperty(PROP_PASSWORD).getValue().toCharArray());
+            }
+
+            synchronized (mqttClientLock) {
+                mqttClient = new MqttClient(broker, clientID, persistence);
+
+                mqttClient.setCallback(new PublishMQTTCallback());
+                getLogger().info("Connecting to broker: " + broker);
+                mqttClient.connect(connOpts);
+            }
+        } catch(MqttException me) {
+            logger.error("Failed to initialize the connection to the  "+me.getMessage());
+        }
+    }
+
+    @OnStopped
+    public void onStop(final ProcessContext context) {
+        try {
+            synchronized (mqttClientLock) {
+                if (mqttClient != null && mqttClient.isConnected()) {
+                    mqttClient.disconnect();
+                }
+            }
+            logger.info("Disconnected the MQTT client.");
+        } catch(MqttException me) {
+            logger.error("Failed when disconnecting the MQTT client.", me);
+        }
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile flowfile = session.get();
+        if (flowfile == null) {
+            return;
+        }
```

Comment:
PublishMQTT won't spin "without any delay (yield)" because it requires input and under normal circumstances will only get triggered by the controller when there are flowfiles to process. This check is more or less a fail-safe. Yielding is for when there is something wrong with the processor for which waiting may help resolve (failing to reconnect). I don't see this as an instance that warrants yielding.